proposals/0048-container-api.md

```diff
@@ -0,0 +1,194 @@
+
+Eventually, we landed to a common and abstract concept called **Pipeline**. A pipeline is an abstract
+class that is responsible to render any kind of Astro route: page, endpoint, redirect, virtual.
+
+Each pipeline inside the Astro codebase (dev, build, SSR and now test) is responsible to collect the important information in different way,
+but eventually each route must be rendered using the same data, which are:
+- **Manifest**: this is a serializable/deserializable version of our "configuration", however it also contains more information, such as renderers, client directives, styles.
+- **Request**: a standard https://developer.mozilla.org/en-US/docs/Web/API/Request object.
+- **RouteData**: a type that identifies a generic Astro route. It contains information such as the type of route, the associated component, etc.
+- **ComponentInstance**: The instance of an Astro component. This is the **compiled** component. Internally, an Astro component undergoes various transformations via Vite, and it eventually gets consumed by our rendering engine (if it's a page).
+
+## Public APIs
+
+- `AstroContainer::create`: creates a new instance of the container.
+- `AstroContainer.renderToString`: renders a component and return a string.
+- `AstroContainer.renderToResponse`: renders a component and returns the `Response` emitted by the rendering phase.
+
+### `create` function
+
+A container is class exposed via the `astro/container` specifier. Users **shouldn't** create a container using the `new` instance, but they should use the static function `create`:
+
+```js
+const container = await AstroContainer.create()
+```
+
+The reason why `create` returns a promise is because internally we will validate and parse the user configuration. As a developer, you will be able to pass the same configuration that Astro uses:
+
+```js
+import astroConfig from "../src/astro.config.mjs";
+
+const container = await AstroContainer.create({
+    astroConfig
+})
+```
+
+The function `create` will throw an error if there's a validation error.
+
+> [!NOTE]
+> If you use a TypeScript file for your configuration, you are responsible for loading and transforming it to JavaScript.
+
+
+The `container` binding will expose a `renderToString` that accepts a Astro component and will return a `string`:
+
+```js
+import astroConfig from "../src/astro.config.mjs";
+
+const container = await AstroContainer.create({
+    astroConfig
+})
+
+const content = await container.renderToString(AstroComponent);
+```
+
+#### Options
+
+It will be possible to tweak the container APIs with options and such. The `.create` function will accept the following interface:
+
+```ts
+type AstroContainerOptions = {
+    streaming: boolean;
+    renderers: AstroRenderer[];
+    astroConfig: AstroUserConfig;
+}
+```
+
+The `astroConfig` object is literally the same object exposed by the `defineConfig`, inside the `astro.config.mjs` file. This very configuration 
```

Comment:
Have you found use-case for passing the astroConfig yet? I'm worried about getting bug reports when people try to pass through `vite` config and other non-supported stuff.

What do you think about instead raising the relevant config values up to the `AstroContainerOptions` level. That would be stuff like `trailingSlash`.

Reply:
> What do you think about instead raising the relevant config values up to the AstroContainerOptions level. That would be stuff like trailingSlash.

I don't think that would work easily. These options, internally, are used to create a manifest. If we pass these options when rendering a component, it would mean generating a new manifest every time we attempt to render a component, and then discard that manifest. We have to be careful not to override the existing manifest, because the manifest is tight to the lifetime of the instance of the container.

Reply:
Not when rendering, `AstroContainerOptions` is the name of the options passed to `AstroContainer.create()`.

Reply:
Oh, I misunderstood what you meant. Yeah we should be able to provide the individual options.